File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int n = fruits.length;
        Map<Integer, Integer>mpp = new HashMap<>();
        int maxi_fruits = 0;
        int i = 0;

        for(int j = 0; j < n; j++){
            
            // Add every element into map
            mpp.put(fruits[j], mpp.getOrDefault(fruits[j], 0) + 1);
        
         // Here the condition we need to shrink the window
            while(mpp.size() > 2){
                
                // Decrement & Remove the first element (ith) from map
                mpp.put(fruits[i], mpp.get(fruits[i]) - 1);
                if(mpp.get(fruits[i]) == 0){
                    mpp.remove(fruits[i]);
                }
                i++;
            }
            maxi_fruits = Math.max(maxi_fruits, j-i+1);
        }
        return maxi_fruits;
    }
}

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        unordered_map<int, int>mpp;
        int n = fruits.size();
        int maxi_fruits = 0, i = 0;

        for(int j = 0; j < n; j++){
            mpp[fruits[j]]++;

            while(mpp.size() > 2){
                mpp[fruits[i]]--;

                if(mpp[fruits[i]] == 0){
                    mpp.erase(fruits[i]);
                }
                i++;
            }
            maxi_fruits = max(maxi_fruits, j-i+1);
        }
        return maxi_fruits;
    }
};